inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_broadcast.cpp

```diff
@@ -29,19 +31,66 @@ void dynamicToStaticShapeBroadcast(std::shared_ptr<ngraph::Node> target) {
     }
 
-    auto dsr = std::make_shared<ngraph::vpu::op::DynamicShapeResolver>(
-            staticShapeBroadcast->output(0), broadcast->input_value(1));
-    dsr->set_friendly_name(broadcast->get_friendly_name());
+    std::shared_ptr<ngraph::Node> dsr;
+
+    if (broadcast->get_broadcast_spec() == ngraph::op::BroadcastType::BIDIRECTIONAL) {
+        const auto inputShape = broadcast->get_input_partial_shape(0);
+
+        const auto targetShape = broadcast->input_value(1).get_node_shared_ptr();
+        const auto shapeType = targetShape->get_element_type();
+
+        const auto inputShapeDimsCount = inputShape.rank().get_length();
+        const auto targetShapeDimsCount = ngraph::shape_size(broadcast->get_input_partial_shape(1).get_shape());
+
+        const auto inputShapeConst = std::make_shared<ngraph::opset5::Constant>(shapeType,
+                                                                                ngraph::Shape{static_cast<size_t>(inputShapeDimsCount)},
+                                                                                inputShape.get_shape());
+
+        const auto maxRankNode = inputShapeDimsCount > targetShapeDimsCount ? inputShapeConst : targetShape;
+        const auto minRankNode = maxRankNode == inputShapeConst ? targetShape : inputShapeConst;
+        const auto maxRank = maxRankNode == inputShapeConst ? inputShapeDimsCount : targetShapeDimsCount;
+        const auto minRank = minRankNode == inputShapeConst ? inputShapeDimsCount : targetShapeDimsCount;
+
+        ngraph::NodeVector dims;
+        for (int i = 0; i < minRank; i++) {
+            const auto minRankDim = std::make_shared<ngraph::opset5::Gather>(minRankNode,
+                                                                             ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {minRank - i - 1}),
+                                                                             ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {0}));
+            const auto maxRankDim = std::make_shared<ngraph::opset5::Gather>(maxRankNode,
+                                                                             ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {maxRank - i - 1}),
+                                                                             ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {0}));
+            dims.push_back(std::make_shared<ngraph::opset5::Maximum>(minRankDim, maxRankDim));
+        }
 
+        for (int i = maxRank - minRank - 1; i >= 0; i--) {
+            dims.push_back(std::make_shared<ngraph::opset5::Gather>(maxRankNode,
+                                                                    ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {i}),
+                                                                    ngraph::opset5::Constant::create(shapeType, ngraph::Shape{1}, {0})));
+        }
+
+        std::reverse(dims.begin(), dims.end());
```

Comment:
Okay, now I'm confused a bit. Code above seems to me as extracting dimensions from shape in rever order, but then you reverse here again. Why is that?

Reply:
In bidirectional mode the shapes may have different sizes and this broadcast rule is right alignment, so I go from the right to the left. I've used `reverse` just because I've used the `push_back`, but I think it's better to insert in the beggining of the vector, so no `reverse` is needed. Done.

Reply:
Since you push at the beginning instead of back + removed reverse at the same time - does it mean you have now different result from previous version?

input: 0 1 2 3

Previously:

* after for-loop: 3 2 1 0
* after reverse: 0 1 2 3

Now:

* after for-loop: 3 2 1 0

Is it expected?

Reply:
No, it's not fully correct. The for-loop itself goes from the end to the begginning of the input.
You're correctly said about the previous version, but now it still goes from the end to the beginning of the input, but each element is inserted into the beginning of the vector, which means that 0 will be processed last and will be first in the result vector, so the output will be the same: 0 1 2 3

Reply:
Yea, you're right. And as I understand it's impossible to go from left to right and push at the back due to operation specification?

Reply:
Heh, I don't agree on this. Going from the beginning to the end and push_back is idiomatic and always easier to reason about.

Reply:
OK, fixed.